### ISSUE TITLE
Power utilization not calculated over cascaded devices/PDU

### ISSUE THREAD

**Issue by A888R** [netbox-community/netbox#3377]
# Environment
* Python version:  3.6.8
* NetBox version:  2.6.1

### Steps to Reproduce
1. create power panel
2. create power feed
3. create 3 devices (example: 'pdu_1level', 'pdu_2level', 'device')
4. connect 'pdu_1level' to created feed (over power port)
5. connect 'pdu_2level' to 'pdu_1level' (over power port + power outlet)
6. connect 'device' to 'pdu_2level'
7. set 'Allocated draw' for 'device' (example: 300)

### Expected Behavior
In 'pdu_2level' view 'Power Utilization -> Allocated' - 300;
In 'pdu_1level' view 'Power Utilization -> Allocated' - 300;
In power feed view 'Utilization (Allocated)' - 300.

### Observed Behavior
In 'pdu_2level' view 'Power Utilization -> Allocated' - 300;
In 'pdu_1level' view 'Power Utilization -> Allocated' - **0**;
In power feed view 'Utilization (Allocated)' - **0**.

**Comment by larsuhartmann** [comment]
Same behaviour on our Side - It would be nice to have this Feature - it is not uncommon to add PDUs to USVs and USVs to Power Feeds especially in smaller Edge Cabinets..

**Comment by roblarose** [comment]
Agreed.  In a "server room" (versus a large datacenter?) we often see multiple PDUs on a single UPS (backed by a single circuit/feed), so at least 3 hops for most equipment.

Feed -> UPS -> PDU -> Device

**Comment by jeremystretch** [comment]
Issues like this illustrate why participation in beta releases is so important. Modifying the current approach to support recursive calculations efficiently would require a substantial amount of work. Do we have any volunteers?

**Comment by larsuhartmann** [comment]
We could switch our Netbox instance to a beta branch to test it out - no problem!

**Comment by jeremystretch** [comment]
@larsuhartmann not beta testing, actually committing to doing the work.

**Comment by andrewcleveland** [comment]
@jeremystretch Is a recursive SQL query the right way to go about this?
[nested_power_draw_legs.patch.txt](https://github.com/netbox-community/netbox/files/3884733/nested_power_draw_legs.patch.txt)
Just a proof of concept (individual legs only).

**Comment by chas0rde** [comment]
This definetly is also an issue for us. Sadly we do not have internal coding-capabilities we could supply but we would be happy if anyone could support this think

**Comment by hSaria** [comment]
Using a recursive query in the database seems like the best approach; as far as I can tell, performing the recursion outside of it would entail O(n) as opposed to O(1) in it.

The work that @andrewcleveland has done looks good, but NetBox is currently prone to loops with that query because the current power model allows for loops, be it direct (power port connected back to a power outlet on the same device) or indirect (looped over multiple devices).

It might be worth addressing the power loop first as it could very well fix this issue in the process.

**Comment by phurrelmann** [comment]
@jeremystretch could you please check if the proposed solution (see @andrewcleveland and #3782) is accetable? I'd really like to have this fixed. We have several cascaded PDUs in our racks and thus currently can't montor the power usage.

**Comment by lps-rocks** [comment]
I'd like to see this implemented. We run a small lab that has UPS devices and PDU devices in each rack - so being able to attribute the power draw all the way back to the UPS / Circuit.

**Comment by hSaria** [comment]
I'll take care of creating the PR for this. I was initially wanting to prevent power loops, but I'm more than fine with keeping them (see #3782 for more info).

Instead, I'll limit the recursive query to a number of iterations (maybe 16) and use `UNION` instead of `UNION ALL` (former doesn't include duplicates). If there is a loop, it'll cause the query to hit the iteration limit, but will still return valid and non-inflated/non-duplicate results.

I'll make sure to put enough tests to detect any regression in this as it involves a raw query.

**Comment by roblarose** [comment]
My hero!

--Rob


*--*

*Rob LaRose*

*engineer | rock paper scissors <https://rockpaperscissors.com/> 245 5th
ave 23rd floor new york ny 10016 | O 212 255 6446*

**Comment by hSaria** [comment]
@andrewcleveland Thanks a lot for your query; I modified it to remove looped entries, but apart from that, it pretty much worked.

Just need to add tests for this and then I'll fire up a PR.

**Comment by phurrelmann** [comment]
@hSaria, thank you. I'm eager to test this :)

**Comment by roblarose** [comment]
Me too!

--Rob


*--*

*Rob LaRose*

*engineer | rock paper scissors <https://rockpaperscissors.com/> 245 5th
ave 23rd floor new york ny 10016 | O 212 255 6446*

**Comment by Blackraz0r** [comment]
Can someone help me with this?
my calculation also dows not work.

Is created a panel wich goes into a feed wich goes into a device called "UPS" wich goes into a device called "PDU" wich goes to other devices.

example:
Panel->Feed->(PowerPort of UPS - Power Outlet of UPS) ->
(PowerPort of PDU - Power Outlet of PDU) -> Switch

the allocated draw of the switch is not in the utilization of the pdu. and also not further down the chain.

can someone eli5 me how to do power in netbox? Have the strange feeling i am missing something here. since english is not my main language i also dont really understand what allocated and maximum draw means. the manual is kinda limited on this.
Like allocated on the pdu is the power the pdu itself neets for running its software (metered with webinterface) and the like ?

**Comment by roblarose** [comment]
Looks like jeremystretch already intercepted this but yeah please don't let this go stale/wontfix.  I'm eagerly awaiting!

**Comment by phurrelmann** [comment]
thank you Jeremy for accepting this one. I'm just spinning up a new vm to test this, as I could not do it yet in our testing-environment

**Comment by phurrelmann** [comment]
I just tested this in a new test setup and it is working great. Cascaded PDU get correctly summed and it now possible to handle the use cases described in this ticket. Thank you very much for your PR @hSaria and @jeremystretch for your tremendous work on this project and the patience with us users.

**Comment by roblarose** [comment]
Is this feature included in the recent 2.7.4 release?  Or still outside the releases?


--Rob

--
Rob LaRose
engineer | rock paper scissors 
245 5th ave 23rd floor new york ny 10016 | O 212 255 6446

**Comment by jeremystretch** [comment]
@roblarose No, it hasn't been implemented yet. I believe @phurrelmann means that he tried out the pull request submitted by @hSaria above.

**Comment by roblarose** [comment]
Roger that.  Thank you.  :-)

--Rob


*--*

*Rob LaRose*

*engineer | rock paper scissors <https://rockpaperscissors.com/> 245 5th
ave 23rd floor new york ny 10016 | O 212 255 6446*

**Comment by phurrelmann** [comment]
Can we get this into 2.8? For me the PR looks solid, but I can't review it :(
@hSaria already put so much effort into it. It would be a pity to let this rot further

**Comment by RHuehne** [comment]
is this topic still being focused by the developers or won't this be fixed? 
Thank you.

**Comment by MarcHagen** [comment]
Please do not close this. still needing this function

**Comment by larsuhartmann** [comment]
Same here! It would be really nice to have this!

**Comment by jeremystretch** [comment]
This has been tagged "needs owner" for several months. If no one is interested in committing to the work, it will be closed. @MarcHagen @larsuhartmann are either of you volunteering to own the implementation?

**Comment by hSaria** [comment]
Hey Jeremy,

I think the issue here is that we don’t have an acceptable solution without changing the existing behavior in power modelling. As you and I discussed in previous PRs, raw SQL is difficult to maintain and calculating a topology tree and storing it as a field in the model is somewhat complex.

Because loops are currently permitted in power modelling, we can’t use MPTT. Otherwise, that’s one solution that might work. If you got something in mind, I can try to help where I can.

Regards,
   Saria

**Comment by MarcHagen** [comment]
i would love to help with this, but my Python skills are just not up to the task, and this is a big "task/issue/feature". So sorry :(

**Comment by larsuhartmann** [comment]
-Is there any way to "sponsor" someone to implement this feature?
- we have the same situation at work: it would be nice to have this feature but none of us have the skills to implement it :-\

**Comment by Starkstromkonsument** [comment]
The function required for the calculation seems to be already implemented at

https://github.com/netbox-community/netbox/blob/b2e05aafc1a99d14fea2ab9c446a8dd7b4cd6375/netbox/dcim/models/device_components.py#L327

It is called by Power Feeds (for Power Utilization)

https://github.com/netbox-community/netbox/blob/872ba89cad976bca84fab1cbc9360ee63e4039eb/netbox/templates/dcim/powerfeed.html#L111

and Racks (for Power Utilization)

https://github.com/netbox-community/netbox/blob/4ce7dfa55edc0ddc1229c22ad837bcb959c4ec60/netbox/templates/dcim/rack.html#L249

There is another call for devices

https://github.com/netbox-community/netbox/blob/872ba89cad976bca84fab1cbc9360ee63e4039eb/netbox/templates/dcim/device.html#L221

but I was not able to figure out for which devices this is displayed in the front end yet.

I opt for reopening #3782 to prevent power loops (I can't see a valuable use case. Electrical power circuits should always be protected by corresponding fuses/switchgears. Building loops or interconnecting different circuits ends up in a mess and can get very dangerous in case of a failure. This is what devices with several PSUs are built for.) and use the existing function `get_power_draw` recursively for power ports connected to power outlets.

But I regret, I don't have sufficient python skills to so either :-(
